### ISSUE TITLE
Can't set Code Location/Scan names

### ISSUE THREAD

**Issue by chrisseto** [blackducksoftware/hub-detect#213]
I'm in the process of creating a TeamCity plugin that configures and runs hub detect.

It appears that code location names are always set to `{dirname ./..}/{Project Name}/{Project Version}{/bom| scan}`.

When TeamCity clones a repository it places it into an arbitrarily named folder, hub-detect then uses this name for the uploaded code location. When the repository is cloned into a different folder, hub-detect does not clear the old code locations because the names no longer match.

I've tried the following settings to see if I could change the behavior:
```
// SPRING_APPLICATION_JSON
{
"detect.project.codelocation.delete.old.names":"true",
"detect.source.path":".",
"detect.project.codelocation.prefix":"JustATest"
}

//Loaded as
sourcePath = <snip>/buildAgent/work/1da1958be684ed96 [.]
projectCodeLocationDeleteOldNames = true
projectCodeLocationPrefix = JustATest
```

Is there a setting that I've glossed over or another work around?

**Comment by jamesrichard91** [comment]
In your case there is not a way to override the name in the way you want. A workaround you could do in the plugin is, you "could" copy the contents of the workspace and put it into a directory you can control the name of and run Detect against that.

**Comment by ekerwin** [comment]
There is no way to completely control the code location name that is generated - at least for now. If you can't control the name of where TeamCity clones the repo, I don't see how, with current detect capabilities, you could do what you are attempting.

**Comment by imbeyondboredom** [comment]
Is that not a bug though? I would think that most build systems clone the project into a random directory.

What utility does the directory name provide? It seems like the directory the file is in is irrelevant to the content or potential security of the files. As seen in this case, the folder name is random. 

When it's not random - I don't see how it would prevent collisions. As an example, lets say that we have a master and develop branch and we have forgotten to change the version number between those two branches. If we move our code down a directory then we'll still get master and develop scans overwriting each other despite them being in different stages of deployment.

**Comment by ekerwin** [comment]
If a project/version can't be determined, the only distinct thing would be the source path.

**Comment by imbeyondboredom** [comment]
Hmm, I could see that. Although wouldn't it be better to fail with an error code in that case as that's a misconfigured run of the script? 

If hub-detect didn't detect my project name correctly I would rather know and specify it instead of relying on people to either have their code in the same directory or in a different one depending on their desired outcome. (whether they want the scans to overwrite themselves)

**Comment by ekerwin** [comment]
That is certainly worth considering. We originally wanted to avoid failures at any cost and if we could find an "acceptable" default, we used it.

Code location names are one of the most confusing concepts for many of our users. Whether or not a particular run of hub-detect will replace BOM entries (if we use the same code location name) or add BOM entries (if we produce a different code location name) is a constant headache for us to support.

It seems like right now we are trying to find a way to support the power users (those who might be commenting on or creating github issues, for example!) as well as the wide audience of users who just invoke hub-detect with the defaults.

If we added a code location name override, our concern has always been that users who aren't clear on its use or consequences would flood us with support requests.

Would you be happy with something like --detect.project.version.names.strict=true? This would omit the folder name part of the code location name if we find a project/version and if we don't find a project/version, it would fail with a new (existing?) exit code.

Full disclosure - we discussed this solution and can't figure out why we didn't do it initially...so there may be reasons that this is not a good idea that we will rediscover in implementation if we go down this road.

**Comment by ekerwin** [comment]
TL; DR, Would you be happy with something like --detect.project.version.names.strict=true? This would omit the folder name part of the code location name if we find a project/version and if we don't find a project/version, it would fail with a new (existing?) exit code.

**Comment by chrisseto** [comment]
How would that handle package managers that don't store information about the project itself? Namely, the majority of golang's vendoring utilities. 
We don't need the ability to control all the code location names, just the one(s) in the root of the source directory. Would it be easier to allow overriding the source directory name?

**Comment by ekerwin** [comment]
In the case where the project/version names can't be found, with that flag set to true you'd have to provide them or get the failure.

As for overriding the source directory name, I have no issue with adding a property for that.

**Comment by akamen** [comment]
@chrisseto  or @imbeyondboredom  would the proposed override property be acceptable?   There are some risks with using it, but if you want us to proceed - let us know please.

**Comment by chrisseto** [comment]
Override the source directory name covers my use case.
What would the risks be?

**Comment by jamesrichard91** [comment]
For example : When Detect creates multiple code locations because there are either several package managers found or there is a multi-module project, the override will be used in creating all of the code locations. So you would run the risk that some of the code location names may end up being the same and only the first one will be uploaded to the Hub.

**Comment by taikuukaits** [comment]
Hi chrisseto, 

We added a new property "detect.code.location.name" which allows you to override what we set for the code location names. Detect can find multiple code locations so the first one has the exact override name and then each subsequent code location has the override plus it's index. So you would see "override" then "override 2" then "override 3". This is in PR #262. 

I believe this accomplishes what you want, if not, just re-open.
